### ISSUE TITLE
400 response for /wsapi/used_address_as_primary when adding primary to secondary 

### ISSUE THREAD

**Issue by jrgm** [mozilla/persona#2840]
This doesn't appear to have any negative effect in the UI, but I see a 400 error calling POST /wsapi/used_address_as_primary

STR
1. With a clean profile, use dev.myfavoritebeer.org to sign up as a secondary account
2. After completing that, sign out of dev.myfavoritebeer.org, and bring up the dialog again
3. Click 'add another email address' and enter someone@eyedee.me
4. Enter a password on eyedee.me and complete the flow

Here's the sequence of wsapi calls start from entry of an eyedee.me address:

```
Time         Secs  Sent Recv       RC  Content-type     URL
00:02:33.169 0.334  820  412  GET  200 application/json https://login.dev.anosrep.org/wsapi/address_info?email=abcabdab%2Bs9%40eyedee.me
00:02:33.506 0.205  818  274  GET  200 application/json https://login.dev.anosrep.org/wsapi/have_email?email=abcabdab%2Bs9%40eyedee.me
00:02:35.304 0.208  818  274  GET  200 application/json https://login.dev.anosrep.org/wsapi/have_email?email=abcabdab%2Bs9%40eyedee.me
00:02:44.339 0.217  791  523  GET  200 application/json https://login.dev.anosrep.org/wsapi/session_context
00:02:44.563 0.206  928  298  POST 400 text/plain       https://login.dev.anosrep.org/wsapi/used_address_as_primary
00:02:44.778 0.209  818  274  GET  200 application/json https://login.dev.anosrep.org/wsapi/have_email?email=abcabdab%2Bs9%40eyedee.me
00:02:44.781 0.248  820  412  GET  200 application/json https://login.dev.anosrep.org/wsapi/address_info?email=abcabdab%2Bs9%40eyedee.me
00:02:45.032 0.213  818  274  GET  200 application/json https://login.dev.anosrep.org/wsapi/have_email?email=abcabdab%2Bs9%40eyedee.me
00:02:46.136 0.269 2561  261  POST 200 application/json https://login.dev.anosrep.org/wsapi/add_email_with_assertion
00:02:51.221 0.261  804  523  GET  200 application/json https://login.dev.anosrep.org/wsapi/session_context
00:02:51.490 0.230  800  330  GET  200 application/json https://login.dev.anosrep.org/wsapi/list_emails
```

The request sends the same cookie that is used for all the other, successful
requests, and POSTs this body as application/json:

```
'{"email":"abcabdab+s9@eyedee.me","csrf":"5NTUC8ld+qKxGA9v5gOuuA=="}'
```

The response returns a body of 'Bad Request: that email does not belong to you'.

**Comment by lloyd** [comment]
Nice find.  I'll look at this in a couple hours.

-- lloyd (thumb-typing)

**Comment by jrgm** [comment]
The error repeats (not surprisingly) each time I add another primary.

/me thanks httpfox for the find ;-)

**Comment by seanmonstar** [comment]
it was designed to not bother the UI if it fell over... but, it also should be working :)

**Comment by jrgm** [comment]
For completeness, I don't see this error when I just sign up with a primary as the sole email address (i.e., not adding a primary). That flow doesn't call /wsapi/used_address_as_primary (sorry, I'm behind on what all the new wsapi stuff is).

**Comment by seanmonstar** [comment]
It first checks if your authenticated, since we require you own the email before changing it's status. If you're not authenticated, then you'll have to call `auth_with_assertion`, and we can change the status then.

**Comment by lloyd** [comment]
Some context: we want to display informational screens to a user the _first time_ they use a "primary email address", one where the email provider supports persona.  In order to implement this, the server must know that a user has used an email address via their provider.  Because there is one case where this information is simply not available (the client gets a cert directly from the email provider and then logs into a site), we added this hook.  Basically, if a user is authenticated to persona, and they have an email address that just transitioned to a primary address, then they get helpful messaging, get directed to their provider inside the dialog, and come back.  At this point they previously did not need to interact further with our servers.  To get transition UX awesome, we added this single hook where the client reports to the server "this email address was just used as a primary email address".

Note in other cases, the client authenticates to persona with an assertion, which we use to indicate that the email address was used as a primary address.  In this case, no explicit api call is needed.

So with that context, the issue here seems to be that we're invoking "used_address_as_primary" in the wrong spot, before the user is authenticated with assertion.  Job seems to be to find where to move that api call.

As it stands, pending testing which verifies that transition screens are displayed once and only once this is an eyesore, but innocuous.

**Comment by seanmonstar** [comment]
@lloyd the network call first checks for authentication, otherwise it doesn't send it: https://github.com/mozilla/browserid/blob/dev/resources/static/common/js/network.js#L743

So, the dialog believes the user is authenticated, and then makes the request, and the server cries 400...

**Comment by lloyd** [comment]
ok.  now that's odd.  I obviously analyzed this wrong.  I'll dig deeper!  (or anyone else is welcome to)

thanks @seanmonstar

**Comment by shane-tomlinson** [comment]
I believe @lloyd may be correct here. add_email_with_assertion is called _after_ the the call to used_address_as_primary. 

This means used_address_as_primary is being called for an address that the user does not yet own.

Since the user is already authenticated to the password level, the authentication check in network.usedAddressAsPrimary proceeds without a problem.

```
*** USER SIGNS IN WITH EYEDEE.ME ***
[09:47:37.897] GET https://eyedee.me/js/sign_in.js [HTTP/1.1 200 OK 407ms]
[09:47:37.898] GET https://login.dev.anosrep.org/include.js [HTTP/1.1 304 Not Modified 103ms]
[09:47:37.975] GET https://login.dev.anosrep.org/authentication_api.js [HTTP/1.1 200 OK 206ms]
[09:47:38.087] GET https://themes.googleusercontent.com/static/fonts/lora/v5/nAKwuw6_dIh5kwvpj3ShNfesZW2xOQ-xsNqO47m55DA.woff [HTTP/1.1 200 OK 148ms]
[09:47:38.088] GET https://themes.googleusercontent.com/static/fonts/opensans/v6/EInbV5DfGHOiMmvb1Xr-hqRDOzjiPcYnFooOUGCOsRk.woff [HTTP/1.1 200 OK 170ms]
[09:47:38.299] GET https://eyedee.me/api/have_user?user=random.email%2B124556 [HTTP/1.1 200 OK 201ms]
--
[09:47:59.884] POST https://eyedee.me/api/signin [HTTP/1.1 200 OK 283ms]

*** USER IS REDIRECTED BACK TO THE DIALOG ***
[09:48:00.157] GET https://login.dev.anosrep.org/sign_in#AUTH_RETURN [HTTP/1.1 304 Not Modified 102ms]
[09:48:00.445] GET https://login.dev.anosrep.org/wsapi/session_context [HTTP/1.1 200 OK 338ms]


*** BAD REQUEST BECAUSE THE USER DOESN'T OWN THE ADDRESS ***
[09:48:00.822] POST https://login.dev.anosrep.org/wsapi/used_address_as_primary [HTTP/1.1 400 Bad Request 201ms]
[09:48:01.033] GET https://login.dev.anosrep.org/wsapi/have_email?email=random.email%2B124556%40eyedee.me [HTTP/1.1 200 OK 205ms]
[09:48:01.035] GET https://login.dev.anosrep.org/wsapi/address_info?email=random.email%2B124556%40eyedee.me [HTTP/1.1 200 OK 279ms]
[09:48:01.320] GET https://eyedee.me/browserid/provision.html [HTTP/1.1 200 OK 206ms]
[09:48:01.530] GET https://eyedee.me/js/jquery.js [HTTP/1.1 304 Not Modified 101ms]
[09:48:01.531] GET https://login.dev.anosrep.org/provisioning_api.js [HTTP/1.1 304 Not Modified 104ms]
[09:48:01.689] GET https://eyedee.me/api/whoami [HTTP/1.1 200 OK 197ms]
[09:48:01.918] POST https://eyedee.me/api/cert_key [HTTP/1.1 200 OK 210ms]

*** THE USER OWNS THE ADDRESS AFTER THIS CALL ***
[09:48:02.188] POST https://login.dev.anosrep.org/wsapi/add_email_with_assertion [HTTP/1.1 200 OK 304ms]
```

**Comment by shane-tomlinson** [comment]
I am labeling this 5\* so that this gets fixed before this heads into stage.

**Comment by jrgm** [comment]
Fixed on an awsbox after merge to train-2012.12.07. We'll get this on stage on Wednesday.
